init.lua

```diff
@@ -89,19 +89,51 @@ function torch.type(obj)
 end
 
 --[[ See if a given object is an instance of the provided torch class. ]]
+local function exactTypeMatch(obj_mt, typeSpecNameString)
+  return obj_mt.__typename == typeSpecNameString
+end
+
+local function partialTypeMatch(obj_mt, typeSpecNameString)
```

Comment:
This looks rather heavy with allocations of new tables etc. I'm used to thinking of torch.isTypeOf as a cheap, almost free, operation. Would it not be cheaper to string.reverse your two strings first, then check for their longest common prefix, then check if there are '.' left after that in one of them (this seems to be essentially what you are checking, thought I might have misunderstood)